### DISCUSSION TITLE
Is it possible to use application/x-scittle with DocumentFragment

### DISCUSSION THREAD

**Discussion by kangaroolab** [borkdude/scittle#92] [General]
in the scenario where I have a bootstrap clojurescript within index.html as application/x-scittle, which load other cljs files to run, e.g.:

```
defonce executed? (atom false))

  (when-not @executed?
    (do
      (let [cljs-files 
              ["cljs/init.cljs"]]
            (doseq [file cljs-files]
              (let [script (js/document.createElement "script")]
                (set! (.-type script) "application/x-scittle")
                (set! (.-src script) file)
                (.appendChild (.-body js/document) script)))
                (-> js/scittle .-core .eval_script_tags))
      (reset! executed? true)))
```

Is it possible to use application/x-scittle with DocumentFragment instead of document so the bootstrap itself doesn't get run by scittle eval_script_tag?

e.g. could it be sth like` (-> js/scittle .-core (.eval_script_tags ADocFragment))`

basically if no parameter provided, eval_script_tags will default to document

**Comment by borkdude** [comment]
@kangaroolab There was a PR here to make eval-script-tags receive varargs of Elements that are script tags. I think that should solve your problem? https://github.com/babashka/scittle/commit/06621b2dabfbe0579ff698f10cc3196b436303be
If so, I can make a release.

Side question: if you're using bootstrap cljs, why are you also using scittle?

**Reply by kangaroolab** [reply]
thanks @borkdude. reading the PR, it seems we will need to provide a collection of scripts to be processed correct? if yes, could we just pass in the parent element like DocumentFragment or we have to construct the collection ourselves?

I am happy to test it if you could provide more info how it should work with an example.

as about bootstrap cljs, that is really just my own file name. so I need scittle to eval it for execution when my webapp is loaded. hope this clarifies and helps.

**Reply by borkdude** [reply]
You need to provide script tag elements separately, so `(scittle.core.eval_script_tags e1 e2 e3)`.

You can get these elements from a `DocumentFragment` with the `.querySelector` method right? And then you could call apply to call eval-script-tags on them.

cc @chr15m

**Reply by chr15m** [reply]
Yep exactly. So you can either evaluate all Scittle scripts on the page with `(scittle.core.eval_script_tags)` or you can eval specific ones with a query selector. You mentioned you have one parent node with several scripts in it so you could do something like this:

```clojure
(apply scittle.core.eval_script_tags
       (js/Array.from
         (parentnode.querySelectorAll
           "script[type='application/x-scittle']")))
```

Of course you can find the scripts and loop through them one by one too:

```clojure
(let [scittle-tags
      (.querySelectorAll
        parentnode
        "script[type='application/x-scittle']")]
  (doseq [tag scittle-tags]
    (js/console.log "Reloading" (aget tag "src"))
    (-> js/scittle
        .-core
        (.eval_script_tags tag))))
```

**Reply by borkdude** [reply]
@chr15m Perhaps we can make eval_script_tags deal with the result of querySelectorAll as well

**Reply by chr15m** [reply]
Will take a look.

**Reply by chr15m** [reply]
@borkdude if I am reading the source right we can just remove the ampersand from this:

```clojure
(defn ^:export eval-script-tags [& script-tags]
  (let [script-tags (or script-tags
                        (.querySelectorAll
                          doc "script[type='application/x-scittle']"))]
    (eval-script-tags* script-tags)))
```

I can't actually remember why I thought it was necessary to spread the args like that. :thinking:

**Reply by chr15m** [reply]
Oh I remember now. I wanted to support the case of passing in a single script tag. Should I special case that? Or just require the user to pass in an Array with one element?

**Reply by borkdude** [reply]
We could just handle both

**Reply by borkdude** [reply]
And check the type of the argument

**Comment by kangaroolab** [comment]
is this part of version 0.6.19? I tested but it didn't seem working. it seems ignoring the additional parameter passed and keep reloading all the script with type='application/x-scittle'

**Reply by chr15m** [reply]
No it's not released on npm yet. I'll try to put a pr together today for the changes discussed above.

**Reply by chr15m** [reply]
This is done and merged so it should go out in the next release.

**Comment by kangaroolab** [comment]
thanks both of you. when can we expect the next release? i am happy to test it when it is available

**Reply by borkdude** [reply]
Released

**Comment by kangaroolab** [comment]
thanks...tested and it works as expected.

**Comment by kangaroolab** [comment]
@borkdude and @chr15m 

I put together a local first web app at https://kangaroolab.github.io/tippy/, which uses scittle to enable notebook experience and it can also be used to extend/customize the app itself. you can try demo.json at https://github.com/kangaroolab/tippy  (see readme) in the app.

would love to get your thought/feedback. I have been using it for programming with clojurescript and some other note taking for myself

**Reply by chr15m** [reply]
Super cool @kangaroolab
